api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java

```diff
@@ -0,0 +1,129 @@
+import org.mockito.MockitoAnnotations;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.config.WorkbenchConfig.FireCloudConfig;
+import org.pmiops.workbench.db.dao.BillingProjectBufferEntryDao;
+import org.pmiops.workbench.db.model.BillingProjectBufferEntry;
+import org.pmiops.workbench.firecloud.FireCloudService;
+import org.pmiops.workbench.test.FakeClock;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.Scope;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+@DataJpaTest
+@Import(LiquibaseAutoConfiguration.class)
+public class BillingProjectBufferServiceTest {
+
+  private static final Instant NOW = Instant.now();
+  private static final FakeClock CLOCK = new FakeClock(NOW, ZoneId.systemDefault());
+
+  private static WorkbenchConfig workbenchConfig;
+
+  @TestConfiguration
+  @Import({})
+  @MockBean({
+      FireCloudService.class
+  })
+  static class Configuration {
+    @Bean
+    @Scope("prototype")
+    public WorkbenchConfig workbenchConfig() {
+      return workbenchConfig;
+    }
+  }
+
+
+  @Autowired
+  private BillingProjectBufferEntryDao billingProjectBufferEntryDao;
+  @Autowired
+  private FireCloudService fireCloudService;
+  @Autowired
+  private Provider<WorkbenchConfig> workbenchConfigProvider;
+
+  private BillingProjectBufferService billingProjectBufferService;
+
+  private final long BUFFER_CAPACITY = 5;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+
+    billingProjectBufferService = new BillingProjectBufferService(
+        billingProjectBufferEntryDao, CLOCK, fireCloudService, workbenchConfigProvider
+    );
+
+    workbenchConfig = new WorkbenchConfig();
+    workbenchConfig.firecloud = new FireCloudConfig();
+    workbenchConfig.firecloud.billingProjectPrefix = "test-prefix";
+    workbenchConfig.firecloud.billingProjectBufferCapacity = (int) BUFFER_CAPACITY;
+  }
+
+  @Test
+  public void fillBuffer() {
+    billingProjectBufferService.bufferBillingProject();
+
+    ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
+    verify(fireCloudService).createAllOfUsBillingProject(captor.capture());
+
+    String billingProjectName = captor.getValue();
+
+    assertThat(billingProjectName).startsWith(workbenchConfig.firecloud.billingProjectPrefix);
+    assertThat(billingProjectBufferEntryDao.findByFireCloudProjectName(billingProjectName).getStatusEnum())
+        .isEqualTo(CREATING);
+  }
+
+  @Test
+  public void fillBuffer_capacity() {
```

Comment:
nit: one case worth covering would be reduction of buffer capacity below the number of buffered projects. I'd probably make this a separate test case